### PR TITLE
3scale Backup Documentation

### PR DESCRIPTION
**Pull request by maneta**
As the title states, in this PR we will describe the 3cale backup procedure

### CODE REVIEW

File: docs/day-2-operations/Backups.md [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+
+1. Copy the Zync Database dump to the zync-database pod
+   ```
+   oc cp ./zync-database-backup.gz $(oc get pods -l 'deploymentConfig=zync-database' -o json | jq '.items[0].metadata.name' -r):/var/lib/pgsql/
+   ```
+2. Decompress the Backup File
+    ```
+    oc rsh $(oc get pods -l 'deploymentConfig=zync-database' -o json | jq -r '.items[0].metadata.name') bash -c 'gzip -d ${HOME}/zync-database-backup.gz'
+    ```
+3. Restore the MySQL DB Backup file
```
Comment (slopezz) [inline docs/day-2-operations/Backups.md]: Zync DB is Postgres not MySQL, isn't it? The command described is already OK (psql)

File: docs/day-2-operations/Backups.md [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+### `system-redis`
+This contains Queues for jobs to be processed in background. These are ephemeral and are deleted once processed.
+
+## Backup Procedures
+
+### `system-mysql`
+
+1. Execute MySQL Backup Command
+   ```
+   oc rsh $(oc get pods -l 'deploymentConfig=system-mysql' -o json | jq -r '.items[0].metadata.name') bash -c 'export MYSQL_PWD=${MYSQL_ROOT_PASSWORD}; mysqldump --single-transaction -hsystem-mysql -uroot system' | gzip > system-mysql-backup.gz
```
Comment (slopezz) [inline docs/day-2-operations/Backups.md:74]: Maybe it should be a good idea to also add `--routines` (which is disabled by default, and would dump any possible stored procedure and function, if exists)  and `--triggers` (which is enabled by default, but by adding it explicitly, it is more clear that triggers are being used)

**Review by slopezz — CHANGES_REQUESTED** [review]

File: docs/day-2-operations/Backups.md (hunk 2 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md:74]: Oops I had the -R on the command line probably lost due some editing. Adding it

Comment (maneta) [reply docs/day-2-operations/Backups.md:74]: corrected on 78554c4d6de0d42bd645abcb22f8d1ce50d9e930

File: docs/day-2-operations/Backups.md (hunk 1 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md]: Corrected on 78554c4d6de0d42bd645abcb22f8d1ce50d9e930

File: docs/day-2-operations/Backups.md [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+   appendonly yes
+   ```
+
+12. Re-deploy `system-redis` to reload the default configurations
+   ```
+   oc rollout latest dc/system-redis
+   ```
+
+## Open Issues
+* what about System services and sphinx (index)?
```
Comment (maneta) [inline docs/day-2-operations/Backups.md]: @andrewdavidmackenzie I don't understand, what is the problem with this? AFAIK sphinx will regenerate the indexes from MySQL as needed, right?

File: docs/day-2-operations/Backups.md [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+* How to handle backup/restore of job queues (of different types). They  can be lost or maybe done twice!
```
Comment (maneta) [inline docs/day-2-operations/Backups.md:320]: I guess this is a question for @3scale/system @3scale/backend teams

File: docs/day-2-operations/Backups.md (hunk 1 above)
Comment (slopezz) [reply docs/day-2-operations/Backups.md]: Cool!

File: docs/day-2-operations/Backups.md (hunk 2 above)
Comment (slopezz) [reply docs/day-2-operations/Backups.md:74]: Cool!

**Review by slopezz — APPROVED** [review]

File: docs/day-2-operations/Backups.md [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+1. Execute MySQL Backup Command
+   ```
+   oc rsh $(oc get pods -l 'deploymentConfig=system-mysql' -o json | jq -r '.items[0].metadata.name') bash -c 'export MYSQL_PWD=${MYSQL_ROOT_PASSWORD}; mysqldump --single-transaction -hsystem-mysql -uroot system' | gzip > system-mysql-backup.gz
+   ```
+
+### `system-storage`
+
+1. Archive the system-storage files to another storage.
+    ```
+    oc rsync $(oc get pods -l 'deploymentConfig=system-app' -o json | jq '.items[0].metadata.name' -r):/opt/system/public/system ./local/dir
```
Comment (orimarti) [inline docs/day-2-operations/Backups.md:82]: This should be something like(think the dir is wrong and in my case when I try it it does it on system-master container and it fails, or maybe is because I'm trying with an old AMP):
```
oc rsync $(oc get pods -l 'deploymentConfig=system-app' -o json | jq '.items[0].metadata.name' -r):/opt/system/public ./local/dir
```

File: docs/day-2-operations/Backups.md [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+3. Restore the MySQL DB Backup file
+   ```
+   oc rsh $(oc get pods -l 'deploymentConfig=system-mysql' -o json | jq -r '.items[0].metadata.name') bash -c 'export MYSQL_PWD=${MYSQL_ROOT_PASSWORD}; mysql -R -hsystem-mysql -uroot system < ${HOME}/system-mysql-backup'
+   ```
+
+### `system-storage`
+
+1. Restore the Backup file to system-storage
+   ```
+   oc rsync ./local/dir $(oc get pods -l 'deploymentConfig=system-app' -o json | jq '.items[0].metadata.name' -r):/opt/system/public/system
```
Comment (orimarti) [inline docs/day-2-operations/Backups.md:136]: Looks like directory here would be wrong too.

File: docs/day-2-operations/Backups.md (hunk 5 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md:82]: @orimarti maybe you are using the s3 version?

File: docs/day-2-operations/Backups.md (hunk 6 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md:136]: s3 deployment?

Comment (orimarti) [reply docs/day-2-operations/Backups.md:136]: aaa true

**Review by orimarti — APPROVED** [review]

File: docs/day-2-operations/Backups.md (hunk 5 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md:82]: Hmmm, Maybe would be good to describe that this procedure justs works/makes sense for the FS deployment. and for s3, maybe would be good to recommend some s3 policy for backup...

I will set an OC Cluster up machine and deploy an RWX 3scale on there, would be useful for restore tests.

File: docs/day-2-operations/Backups.md [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+The backup and restore procedures for persistent data are required to vary depending on the storage used, to ensure the backups and restores preserve data conistsency (e.g. that a partial write, or a partial transaction is not captured). i.e. It is not sufficient to backup the underlying Persistent Volumes for a database, but instead the databases backup mechanisms should be used.
+
+Also, some parts of the data are synchronized between different components. One copy is considered the "source of truth" for the data set, and the other a copy that is not modified locally, just synchronized from the "source of truth". In these cases, upon restore, the "source of truth" should be restored and then the copies in other components synchronized from it.
+
+## Data Sets
+This section goes into more detail on the different data sets in the different persistent stores, their purposes, the storage type used and if it is the "source of truth" or not.
+
+The full state of a 3scale deployment is stored across these services and their persistent volumes:
+
+* `system-mysql`:  Mysql database (`mysql-storage`)
```
Comment (hallelujah) [inline docs/day-2-operations/Backups.md:28]: Should we consider other supported database too? i.e. Oracle for now

File: docs/day-2-operations/Backups.md [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+* `system-storage`: Volume for Files
+* `zync-database`: Postgres database for `zync` component. This uses "HostPath" as storage. If the pod is moved into another node the data is lost. That is not a problem because the data are sync jobs and does not need to be 100% persistent.
```
Comment (hallelujah) [inline docs/day-2-operations/Backups.md:30]: When the job has not run yet and the database is lost, then the sync will not happen unless an update is made on `system` side.

So losing data may lead to lost symchronization. The way to recover would be for `system` to trigger the synchronization like for `backend` sync

File: docs/day-2-operations/Backups.md [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+This is for storing files to be read and written by the `System` component. They fall into two categories:
+
+* Configuration files read by the System component at run-time
+* Static files (HTML, CSS, JS, etc) uploaded to System by it's CMS feature, for the purpose of creating a Developer Portal.
+
+Note that `System` can be scaled horizontally with multiple pods uploading and reading said static files, hence the need for a `RWX` PersistentVolume.
+
+### `zync-database`
+This is a relational database for storing information related to the synchronization of identities between 3scale and an Identity provider.
+This information is not duplicated in other components and this is the sole source of truth.
```
Comment (hallelujah) [inline docs/day-2-operations/Backups.md:52]: That's correct. Though `system-database` has some events logging table which could be used to synchronize zync events, thus reconstructing partially the `zync-database`

File: docs/day-2-operations/Backups.md [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+### `backend-redis`
+This contains multiple data sets used by the `Backend` component:
+
+* Usages: This is API usage information aggregated by `Backend`. It is used by `Backend` for rate-limiting decisions and by `System` to display analytics information in the UI or via API.
+* Config: This is configuration information about Services, Rate-limits, etc that is synchronized from `System` via an internal API. This is NOT the source of truth of this info, `System` and `system-mysql` is.
+* AuthKeys: Storage of OAuth keys created directly in `Backend`. This is the source of truth for this information.
+* Queues: Queues of Background Jobs to be executed by worker processses. These are ephemeral and are deleted once processed.
+
+### `system-redis`
+This contains Queues for jobs to be processed in background. These are ephemeral and are deleted once processed.
```
Comment (hallelujah) [inline docs/day-2-operations/Backups.md:65]: Some jobs are not duplicated in other components. Making this database its sole source of truth.

File: docs/day-2-operations/Backups.md (hunk 3 above)
Comment (hallelujah) [reply docs/day-2-operations/Backups.md]: Yes sphinx will regenerate indices from `system-database`
No need to backup sphinx indices

File: docs/day-2-operations/Backups.md (hunk 7 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md:28]: @hallelujah I'm basing myself on the productized provided templates: 

- https://github.com/3scale/3scale-amp-openshift-templates/tree/master/amp

As Oracle is proprietary software, I guess that the Oracle should provide support for backup and restore of their database over Openshift. 

But I guess we need to find some way to communicate that.

File: docs/day-2-operations/Backups.md (hunk 4 above)
Comment (hallelujah) [reply docs/day-2-operations/Backups.md:320]: So for system: some jobs are only stored in system-redis but we could use event logging to replay them
For now there is no `replay` feature in `system` and not all the jobs are triggered by an event, so they cannot be replayed, but we could work on that.

However the `replay` will only work for events no older than 1 week see: https://github.com/3scale/porta/blob/ba76f595b24792f40292bf04e566087b3c75593f/app/workers/purge_old_events_worker.rb#L7
and
https://github.com/3scale/porta/blob/ba76f595b24792f40292bf04e566087b3c75593f/app/lib/event_store/event.rb#L59-L61


Doing them twice is a real issue like sending emails or charging credit card ... No answer for that for now

File: docs/day-2-operations/Backups.md (hunk 7 above)
Comment (hallelujah) [reply docs/day-2-operations/Backups.md:28]: @andrewdavidmackenzie thoughts about that?

File: docs/day-2-operations/Backups.md (hunk 8 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md:30]: @hallelujah so you think we should make this storage persistent and move away from `emptyDir` storage?

File: docs/day-2-operations/Backups.md (hunk 10 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md:65]: So backup is needed right? The restore could be tricky and lead to inconsistency @hallelujah ?

File: docs/day-2-operations/Backups.md (hunk 4 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md:320]: @hallelujah any ideas on how could we solve this?

File: docs/day-2-operations/Backups.md [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+This information is not duplicated in other components and this is the sole source of truth.
+
+### `backend-redis`
+This contains multiple data sets used by the `Backend` component:
+
+* Usages: This is API usage information aggregated by `Backend`. It is used by `Backend` for rate-limiting decisions and by `System` to display analytics information in the UI or via API.
+* Config: This is configuration information about Services, Rate-limits, etc that is synchronized from `System` via an internal API. This is NOT the source of truth of this info, `System` and `system-mysql` is.
+* AuthKeys: Storage of OAuth keys created directly in `Backend`. This is the source of truth for this information.
+* Queues: Queues of Background Jobs to be executed by worker processses. These are ephemeral and are deleted once processed.
+
```
Comment (davidor) [inline docs/day-2-operations/Backups.md:62]: There are others like alerts and integration errors. Backend is also the source of truth for those.
In the past we had "latest transactions" as well, but I think that's deprecated now.

File: docs/day-2-operations/Backups.md (hunk 8 above)
Comment (hallelujah) [reply docs/day-2-operations/Backups.md:30]: Well not necessarily. If system can replay/resync events then it is ok

File: docs/day-2-operations/Backups.md (hunk 4 above)
Comment (davidor) [reply docs/day-2-operations/Backups.md:320]: For backend:

Restoring queues could be a problem because some jobs would be run twice.
Queues are supposed to be empty most of the time. Otherwise it means that reports are taking too long to be effective and rate-limit is not accurate. I think that for that reason losing the last few jobs that got enqueued might be better than restoring a backup with some old jobs that already got run.

File: docs/day-2-operations/Backups.md (hunk 10 above)
Comment (hallelujah) [reply docs/day-2-operations/Backups.md:65]: Yes this needs backup. Inconsistencies could happen like sending email twice on restore.
I just verified, billing jobs do not lead to inconsistencies.

File: docs/day-2-operations/Backups.md (hunk 4 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md:320]: @davidor I have been looking for an option for generating RDB files for just one of the dbs, but looks like it is an issue that the Redis team are not looking forward to addressing (https://github.com/antirez/redis/issues/5369). 

TL;DR: It is impossible today to generate a backup file per DB in Redis. The recommendation is to have different instances of Redis for each database.

To fix this, we should have two different Redis for backend, which means an architecture change on the proposed templates.

@andrewdavidmackenzie WDYT?

File: docs/day-2-operations/Backups.md (hunk 3 above)
Comment (andrewdavidmackenzie) [reply docs/day-2-operations/Backups.md]: But will we need some procedure to re-trigger indexing, after a restore of the DB contents?

For system services, we will need to do anything special on restore? Will depend on the documented procedure for restoring the redis with the job queues that system-services executes?

File: docs/day-2-operations/Backups.md (hunk 4 above)
Comment (andrewdavidmackenzie) [reply docs/day-2-operations/Backups.md:320]: OK, recommendation and David's advice sounds sound.

Not clear what to do yet about System.....?

File: docs/day-2-operations/Backups.md (hunk 7 above)
Comment (andrewdavidmackenzie) [reply docs/day-2-operations/Backups.md:28]: Only if something particular about the DB changes the overall strategy. Then (for me) would be acceptable to do "do a similar procedure for the other Supported DBs, using the documentation from the DB provider" and *maybe* link to their detailed docs. Avoid reproducing the DB docs here, but help the user find them.

File: docs/day-2-operations/Backups.md (hunk 8 above)
Comment (andrewdavidmackenzie) [reply docs/day-2-operations/Backups.md:30]: Not sure System will have the information to know all the syncs required via Zync?
If it does, then sure, I agree...but we need a way in system to force this resync.

On the other hand, having a backup of this DB, and worst case doing the twice, would not be a problem. 

When we have System using Postgres, what storage does it need?
Will need to be persisted and backedup, no?
Then it would make sense to just have Zync use the same Postgres as System, and use the same back-up/restore procedures. If a few syncs get repeated, no problem, no?

File: docs/day-2-operations/Backups.md (hunk 9 above)
Comment (andrewdavidmackenzie) [reply docs/day-2-operations/Backups.md:52]: OK, related to above comment I think. But that maybe useful information to add in this description somewhere....so people know.

File: docs/day-2-operations/Backups.md (hunk 10 above)
Comment (andrewdavidmackenzie) [reply docs/day-2-operations/Backups.md:65]: "sole source of truth" .... for pending jobs, right?
Once executed, they are removed I understand.

So, we go with the backup of it, and risk that some tasks could be done twice?
If there is an enhancement needed here to have System avoid that (maybe for some jobs more important than others) then let's create the JIRA issue for that for the future.

File: docs/day-2-operations/Backups.md (hunk 11 above)
Comment (andrewdavidmackenzie) [reply docs/day-2-operations/Backups.md:62]: Understood. Good points.

But overall recommendation (I think) is:
- restore it
- then sync from system which will over-write some info, others not

and that would take care of the other cases you point out @davidor ?

**Comment by andrewdavidmackenzie** [comment]
Reviewed comments from you and others @maneta Now to review overall text.

File: docs/day-2-operations/Backups.md (hunk 11 above)
Comment (andrewdavidmackenzie) [reply docs/day-2-operations/Backups.md:62]: Add here the missing ones...

File: docs/day-2-operations/Backups.md [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+2. Decompress the Backup File
+    ```
+    oc rsh $(oc get pods -l 'deploymentConfig=zync-database' -o json | jq -r '.items[0].metadata.name') bash -c 'gzip -d ${HOME}/zync-database-backup.gz'
+    ```
+3. Restore the PostgreSQL DB Backup file
+   ```
+   oc rsh $(oc get pods -l 'deploymentConfig=zync-database' -o json | jq -r '.items[0].metadata.name') bash -c 'psql -f ${HOME}/zync-database-backup'
+   ```
+### `backend-redis`
+* After restoring `backend-redis` a sync of the Config information from `System` should be forced, to ensure the information in `Backend` is consistent with that in `System` (the source of truth).
```
Comment (andrewdavidmackenzie) [inline docs/day-2-operations/Backups.md]: Document the command to restore it also, first?

File: docs/day-2-operations/Backups.md [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+   oc rsh $(oc get pods -l 'deploymentConfig=system-redis' -o json | jq '.items[0].metadata.name' -r) bash -c 'mv ${HOME}/data/appendonly.aof ${HOME}/data/appendonly.aof-old'
+   ```
+
+7. Move the Backup file to the POD
+
+   ```
+   oc cp ./backend-redis-dump.rdb $(oc get pods -l 'deploymentConfig=backend-redis' -o json | jq '.items[0].metadata.name' -r):/var/lib/redis/data/dump.rdb
+   ```
+
+8. Re-deploy `system-redis` to load the backup
```
Comment (andrewdavidmackenzie) [inline docs/day-2-operations/Backups.md]: this should be 'backend-redis' no?

File: docs/day-2-operations/Backups.md [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+   ```
+
+5. Rename the `dump.rb` file
+   ```
+   oc rsh $(oc get pods -l 'deploymentConfig=backend-redis' -o json | jq '.items[0].metadata.name' -r) bash -c 'mv ${HOME}/data/dump.rdb ${HOME}/data/dump.rdb-old'
+   ```
+
+6. Rename the `appendonly.aof` file
+   ```
+   oc rsh $(oc get pods -l 'deploymentConfig=system-redis' -o json | jq '.items[0].metadata.name' -r) bash -c 'mv ${HOME}/data/appendonly.aof ${HOME}/data/appendonly.aof-old'
```
Comment (andrewdavidmackenzie) [inline docs/day-2-operations/Backups.md]: the command references 'system-redis'

File: docs/day-2-operations/Backups.md (hunk 12 above)
Comment (andrewdavidmackenzie) [inline docs/day-2-operations/Backups.md:161]: Where is the forcing of the sync described?

**Comment by andrewdavidmackenzie** [comment]
Done for me @maneta !

File: docs/day-2-operations/Backups.md (hunk 4 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md:320]: @andrewdavidmackenzie, as Backend uses the same redis, we can't backup/restore just one of the DBs. So we can't follow David's advise (that's the issue). So what do you think we should do?

File: docs/day-2-operations/Backups.md (hunk 7 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md:28]: Ok, I will try to look for Oracle Containerized DB backup documentation. (Not sure If it is public though.)

File: docs/day-2-operations/Backups.md (hunk 4 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md:320]: @davidor There is any way to minimize the restoration of the queues in the current architecture?

File: docs/day-2-operations/Backups.md (hunk 13 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md]: fixed on 580a5f5dca8d02c8d085be2c5ab700ff53cb7e89

File: docs/day-2-operations/Backups.md (hunk 14 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md]: fixed on 580a5f5dca8d02c8d085be2c5ab700ff53cb7e89

File: docs/day-2-operations/Backups.md (hunk 12 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md:161]: @hallelujah how should we do that?

File: docs/day-2-operations/Backups.md (hunk 7 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md:28]: Guys I was reviewing how we support Oracle in 3scale, and what we do is ask for an Oracle database endpoint and create the system Image with the required Oracle Client files. We don't provide a template or a DeployConfig for Oracle system.

This means that we do not control or provide a recommendation for an environment where the Oracle Database will be deployed. So I think it does make sense to provide backup documentation for this, as the deployment environment could range from  AIX to Windows...

File: docs/day-2-operations/Backups.md (hunk 3 above)
Comment (hallelujah) [reply docs/day-2-operations/Backups.md]: Deploy of system-sphinx pod will do it automatically
https://github.com/3scale/3scale-amp-openshift-templates/blob/a5699103fe5a23818ddca52a18c9d862fabbb7d8/amp/amp.yml#L2903

https://github.com/3scale/porta/blob/master/lib/tasks/openshift.rake#L37

File: docs/day-2-operations/Backups.md (hunk 4 above)
Comment (hallelujah) [reply docs/day-2-operations/Backups.md:320]: Job queues will be lost if the system redis was not backed up. And there is no way to restore most of them. Except from state synchronization between System and Backend.

A redeploy of the system-app pod will resync states of objects between system and backend through the post deploy pod https://github.com/3scale/porta/blob/8d75e618eff1f3050a9fc0f77ecdd576503ee0a7/lib/tasks/openshift.rake#L33

File: docs/day-2-operations/Backups.md (hunk 8 above)
Comment (hallelujah) [reply docs/day-2-operations/Backups.md:30]: If System uses Postgres, it needs to be persistent storage. So zync will benefit of it if it is using the same database. All queues and states will be already persisted, and backup will be like any other postgres database

In case of different databases, then the only way for System to re-trigger the sync is the ability to replay the events that we store (for a short amount of time) in `event_store_events` table.
But this **replay** feature is not done.

File: docs/day-2-operations/Backups.md (hunk 10 above)
Comment (hallelujah) [reply docs/day-2-operations/Backups.md:65]: Yes when a job is executed then it is removed. It would be extremely difficult to control that kind of Job executing only once due to the parallelism execution of those jobs.

How job executing twice could happen is backup and restore while sidekiq services are still running.

File: docs/day-2-operations/Backups.md (hunk 3 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md]: ok @hallelujah so in that sense need to document that in case of system DB Restore we would need to redeploy `system-sphinx` to regenerate the indexes right?

File: docs/day-2-operations/Backups.md (hunk 8 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md:30]: @andrewdavidmackenzie @hallelujah, in this case, could we resolve this conversation?

File: docs/day-2-operations/Backups.md (hunk 9 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md:52]: @hallelujah could you provide some snippet on how to describe this?

File: docs/day-2-operations/Backups.md (hunk 10 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md:65]: @hallelujah so for my understanding there is no need for a jira issue to treat this, as it will be "extremely difficult" right?

In that case, could we resolve this conversation @andrewdavidmackenzie ?

File: docs/day-2-operations/Backups.md (hunk 11 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md:62]: @davidor could you ad the missing ones as @andrewdavidmackenzie requested?

File: docs/day-2-operations/Backups.md (hunk 8 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md:30]: @hallelujah do we have an issue for this replay feature?

File: docs/day-2-operations/Backups.md (hunk 11 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md]: Issue created to address the split of the backend databases: https://issues.jboss.org/browse/THREESCALE-2745

File: docs/day-2-operations/Backups.md (hunk 4 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md:320]: Issue created to address the split of the backend databases: https://issues.jboss.org/browse/THREESCALE-2745

File: docs/day-2-operations/Backups.md (hunk 2 above)
Comment (ciaran-byrne) [inline docs/day-2-operations/Backups.md:74]: You may wish to date stamp these output files, if backups are to be run regularly

File: docs/day-2-operations/Backups.md [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+1. Archive the system-storage files to another storage.
+    ```
+    oc rsync $(oc get pods -l 'deploymentConfig=system-app' -o json | jq '.items[0].metadata.name' -r):/opt/system/public/system ./local/dir
+    ```
+
+### `zync-database`
+
+1. Execute Postgres Backup Command
+   ```
+    oc rsh $(oc get pods -l 'deploymentConfig=zync-database' -o json | jq '.items[0].metadata.name' -r) bash -c 'pg_dumpall -c --if-exists' | gzip > zync-database-backup.gz
```
Comment (ciaran-byrne) [inline docs/day-2-operations/Backups.md:90]: You may wish to date stamp these output files, if backups are to be run regularly

File: docs/day-2-operations/Backups.md [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+    ```
+
+### `backend-redis`
+
+1. Backup the dump.rb file from redis
+   ```
+   oc cp $(oc get pods -l 'deploymentConfig=backend-redis' -o json | jq '.items[0].metadata.name' -r):/var/lib/redis/data/dump.rdb ./backend-redis-dump.rdb
```
Comment (ciaran-byrne) [inline docs/day-2-operations/Backups.md:98]: You may wish to date stamp these output files, if backups are to be run regularly

File: docs/day-2-operations/Backups.md [hunk 17]
```diff
@@ -0,0 +1,272 @@
...
+   ```
+
+### `system-redis`
+
+1. Backup the dump.rb file from redis
+
+   ```
+   oc cp $(oc get pods -l 'deploymentConfig=system-redis' -o json | jq '.items[0].metadata.name' -r):/var/lib/redis/data/dump.rdb ./system-redis-dump.rdb
```
Comment (ciaran-byrne) [inline docs/day-2-operations/Backups.md:106]: You may wish to date stamp these output files, if backups are to be run regularly

File: docs/day-2-operations/Backups.md [hunk 18]
```diff
@@ -0,0 +1,272 @@
...
+   ```
+### `backend-redis`
+* After restoring `backend-redis` a sync of the Config information from `System` should be forced, to ensure the information in `Backend` is consistent with that in `System` (the source of truth).
+
+1. Edit the `redis-config` configmap
+   ```
+   oc edit configmap redis-config
+   ```
+
+2. Comment `SAVE` comands in the `redis-config` configmap
```
Comment (ciaran-byrne) [inline docs/day-2-operations/Backups.md]: This markdown does not render correctly

File: docs/day-2-operations/Backups.md [hunk 19]
```diff
@@ -0,0 +1,272 @@
...
+   oc rollout latest dc/backend-redis
+   ```
+
+9. Edit the `redis-config` configmap
+   ```
+   oc edit configmap redis-config
+   ```
+
+10. Uncomment `SAVE` comands in the `redis-config` configmap
+   ```
```
Comment (ciaran-byrne) [inline docs/day-2-operations/Backups.md]: This markdown does not render correctly

**Review by ciaran-byrne — CHANGES_REQUESTED** [review]

File: docs/day-2-operations/Backups.md (hunk 2 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md:74]: Sure, but this aims to be general documentation, more like a guide on how to backup 3scale (for clients), so I was trying to be more generic an straight forward as possible.

 @fbolton How do other products treat this kind of documentation? Do you think it would be useful to add timestamps?

File: docs/day-2-operations/Backups.md (hunk 19 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md]: corrected on 9391f8d172fbe61a8c6f0a1eafba2452a19aa904

File: docs/day-2-operations/Backups.md (hunk 18 above)
Comment (maneta) [reply docs/day-2-operations/Backups.md]: corrected on 9391f8d172fbe61a8c6f0a1eafba2452a19aa904
